pyqtgraph/widgets/MultiAxisPlotWidget.py

```diff
@@ -0,0 +1,427 @@
+        plotitem.addItem(chart)
+        # keep plotitem inside chart
+        chart.plotItem = plotitem
+        # keeptrack of connections
+        if chart.name not in self._signalConnectionsByChart:
+            self._signalConnectionsByChart[chart.name] = {}
+        # keep axis track
+        chart.axes = [x_axis, y_axis]
+        # keep chart
+        self.charts[name] = chart
+        # create a mapping for this chart and his axis
+        self.axes[x_axis].charts.append(name)
+        self.axes[y_axis].charts.append(name)
+        return chart, plotitem
+
+    def clearLayout(self):
+        """Clears the widget but keeps all created axis and charts."""
+        while self.layout.count() > 0:
+            item = self.layout.itemAt(0)
+            self.layout.removeAt(0)
+            self.scene().removeItem(item)
+            del item
+        for chart_name, chart in self.charts.items():
+            self.disconnect_all(chart)
+
+    def makeLayout(self, axes=None, charts=None):
+        """Adds all given axes and charts to the widget.
+
+        Parameters
+        ----------
+        axes : list of str or None:
+            The names associated with the axes to show.
+        charts : list of PlotItems or None:
+            The names associated with the charts to show.
+        """
+        self.clearLayout()
+        shown_axes = self.show_axes(axes)
+        shown_charts = self.show_charts(charts)
+        if len(shown_charts) != 0:
+            top_level = shown_charts[list(shown_charts)[-1]]
+            # FROM "https://github.com/pyqtgraph/pyqtgraph/pull/2010" by herodotus77
+            # propagate mouse actions to charts "hidden" behind
+            for k, c in self.charts.items():
+                self.connect_signals(top_level, c)
+        # MOVE LEGEND TO LAYOUT
+        if self.pi.legend is not None:
+            self.pi.legend.setParentItem(self.pi)
+        self.update()
+
+    def show_axes(self, axes=None):
+        """Shows all the selected axes."""
+        # SELECT AND ASSEMBLE AXES
+        if axes is None:
+            axes = list(self.axes)
+        lo = {
+            "left": [],
+            "right": [],
+            "top": [],
+            "bottom": [],
+        }
+        for k, a in self.axes.items():
+            if k in axes:
+                lo[a.orientation].append(a)
+        vx = len(lo["left"])
+        vy = 1 + len(lo["top"])
+        # ADD TITLE ON TOP
+        self.pi.titleLabel.show()
+        self.layout.addItem(self.pi.titleLabel, 0, vx)
+        # ADD MAIN PLOTITEM
+        self.vb.show()
+        self.layout.addItem(self.vb, vy, vx)
+        # ADD AXIS
+        shown_axes = {"x": {}, "y": {}}
+        for x, a in enumerate(lo["left"] + [None] + lo["right"]):
+            if a is not None:
+                a.show()
+                self.layout.addItem(a, vy, x)
+                shown_axes["x"][a.name] = a
+        for y, a in enumerate([None] + lo["top"] + [None] + lo["bottom"]):
```

Comment:
slight simplification...

```python
for y, a in enumerate(lo["top"] + [None] + lo["bottom], start=1)
```

I may actually prefer for something like this to do 

```python
for y, a in zip(itertools.count(1, step=2), lo["top"] + lo["bottom"]):
```

and forget about inserting None's into the list.  I'll leave it up to you which you prefer.

Reply:
i prefer not to import itertools, used the first one